Box_intersection_d/include/CGAL/Box_intersection_d/segment_tree.h

```diff
@@ -308,6 +308,56 @@ void dump_box_numbers( ForwardIter begin, ForwardIter end, Traits /* traits */ )
     std::cout << std::endl;
 }
 
+
+template<class T>
+class Has_member_report
+{
+private:
+  template<class U, U>
+  class check {};
+
+  template<class C>
+  static char f(check<bool(C::*)(int), &C::report>*);
+
+  template<class C>
+  static int f(...);
+public:
+  static const bool value = (sizeof(f<T>(0)) == sizeof(char));
+};
```

Comment:
Can you please make it more robust then. I took it from the existing `Has_something` classes, so either they need less sophistication, or they have the same problem.

Reply:
> Can you please make it more robust then. I took it from the existing `Has_something` classes, so either they need less sophistication, or they have the same problem.

They all have the same problem: if anything goes wrong in the detection, then at compile-time the compiler falls back to the other strategy (usually the simplest one, that does not use the member function), without any error or warning.

As we have seen in the other PR (https://github.com/CGAL/cgal/pull/6412/commits/a1b8f81af6df406ac327d674e916fa00d7b10983), anything that makes the real member function different from what is detected can be a problem:
  - a typo in the name,
  - different constness of the member function,
  - different constness of the parameters,
  - different types of parameter that could work because of implicit conversions (`unsigned int`, `std::size_t`, `int`, for example).

I do not know if that can be made robust with C++14: one would have to think about all the different possibilities and add them in the overload set of `f` in `Has_member_report`. C++20 concepts are clearly a possible solution for that, though. So at least it could be made robust for users of C++20 or later.

So, with C++14, I would say: use that scarcely, only when that is actually required, because that sort of imperfect smart trick has a cost: possible future bugs, and maintenance cost (because any chance we make will be lest undetected by the testsuite, even if something goes wrong).

Reply:
@mglisse I would like to invoke you, there. Maybe you have a different perspective from mine. Actually, I would really like if you had a different view on that, either in the direction of robustifying the member functions detection, or in my advice of using that trick scarcely.

Reply:
I want to add that `Has_member_report` will probably lead to a compilation error if `T::report` is an overload set instead of a member function.

Reply:
@mglisse Actually, I think there is a way to make the member detection acts more like C++20, concepts: https://stackoverflow.com/a/62292282/1728537. I do not understand why the author of that answer says it requires C++14.

Reply:
Hmm, why not use expression SFINAE?

Reply:
> Hmm, why not use expression SFINAE?

That right. As far as I know, that is what is used in the answer https://stackoverflow.com/a/62292282/1728537, and I do not get why the poster of that answer speak about C++17. As far as I know, that is C++11. Oh, yes I know: that is to implement the generic member detector, using C++17 `constexpr` lambda expressions. But indeed we should use expression SFINAE, to get the detection more robust.

Reply:
@lrineau can you please just do what you propose, at least for one function so that we have an example.

Reply:
@afabri  you could also add a Struct flag indicating that the method are present in the class. Checking for that flag will not depend on the function signature.

Reply:
> @lrineau can you please just do what you propose, at least for one function so that we have an example.

I have implemented that in commit https://github.com/CGAL/cgal/pull/6461/commits/6260d5a4e47f891deb6cd8eed47d4b73a88ae884 along with a test file.

Reply:
I want to see the results in the testsuite, first.

Reply:
Also I had asked for an example only. Forgot to write "thanks"too.  Thanks, But you confirm that those should then also use your approach?

Reply:
I think that is indeed more robust: that tests if the expression `x.report(0) == true` is valid. That addresses most of the concerns I told about in https://github.com/CGAL/cgal/pull/6461#discussion_r839550495.

I also think the proper testing is very important, to make sure the code is robust and survives future refactoring. In commit https://github.com/CGAL/cgal/commit/6260d5a4e47f891deb6cd8eed47d4b73a88ae884 I have added a test, that both tests positive and negative results with static assertions.